specs/eip4844/polynomial-commitments.md

```diff
@@ -226,12 +301,64 @@ def compute_kzg_proof(polynomial: Sequence[BLSFieldElement], z: BLSFieldElement)
     return KZGProof(g1_lincomb(bit_reversal_permutation(KZG_SETUP_LAGRANGE), quotient_polynomial))
 ```
 
+#### `compute_aggregated_poly_and_commitment`
+
+```python
+def compute_aggregated_poly_and_commitment(
+        blobs: Sequence[Blob],
+        kzg_commitments: Sequence[KZGCommitment]) -> Tuple[Polynomial, KZGCommitment, bytes]:
+    """
+    Return the aggregated polynomial and aggregated KZG commitment.
+    """
+    # Generate random linear combination challenges
+    r, h = hash_to_bls_field(DOMAIN_SEPARATOR_AGGREGATE_PROTOCOL, blobs, kzg_commitments)
+    r_powers = compute_powers(r, len(kzg_commitments))
+
+    # Create aggregated polynomial in evaluation form
+    aggregated_poly = Polynomial(poly_lincomb([blob_to_field_elements(blob) for blob in blobs], r_powers))
+
+    # Compute commitment to aggregated polynomial
+    aggregated_poly_commitment = KZGCommitment(g1_lincomb(kzg_commitments, r_powers))
+
+    return aggregated_poly, aggregated_poly_commitment, h
+```
+
+#### `compute_aggregate_kzg_proof`
+
+```python
+def compute_aggregate_kzg_proof(blobs: Sequence[Blob]) -> KZGProof:
+    commitments = [blob_to_kzg_commitment(blob) for blob in blobs]
+    aggregated_poly, aggregated_poly_commitment, h = compute_aggregated_poly_and_commitment(blobs, commitments)
+    x, _ = hash_to_bls_field(DOMAIN_SEPARATOR_EVAL_PROTOCOL + h, [aggregated_poly], [aggregated_poly_commitment])
```

Comment:
small nit: 

`h + DOMAIN_SEPARATOR_EVAL_PROTOCOL` is more natural than `DOMAIN_SEPARATOR_EVAL_PROTOCOL + h` because the previous transcript state `h` may already be saved in the hash buffer and so starting a new protocol would just mean we append `DOMAIN_SEPARATOR_EVAL_PROTOCOL` to the buffer

Reply:
I think it should be the other way around:
1. I think the domain separator should be at the very start of the hash function, otherwise it does not prevent random oracle values from colliding.
2. h is not the full transcript, but only the output of the hash, so this is trivial for efficiency purposes.

Reply:
Switching h to be appended first instead of second, should not change random oracle collisions I think, unless the hash function is bad.

Instead of thinking of putting it at the beginning of the hash function, I'm thinking of putting it at the beginning of the sub protocol. Ie each new protocol if given a unique domain separator will effectively have a "different" hash function or RO

Reply:
I do not believe that what you suggest functions as a domain separator. If your hashed string is
`"DOMAIN_SEPARATOR_PROTOCOL1" + TRANSCRIPT_PROTOCOL1 + "DOMAIN_SEPARATOR_PROTOCOL2" + TRANSCRIPT_PROTOCOL2`,
then what guarantees you that `"DOMAIN_SEPARATOR_PROTOCOL2"` does not appear somewhere in `TRANSCRIPT_PROTOCOL1`?